core/trie/proof.go

```diff
@@ -54,623 +55,586 @@ func (e *Edge) Len() uint8 {
-		} else if sNodeEdge == nil && !isLeaf { // Internal Binary
-			proofNodes = append(proofNodes, sNodeBinary)
-		} else if sNodeEdge != nil && isLeaf { // Leaf Edge
-			proofNodes = append(proofNodes, sNodeEdge)
-		} else if sNodeEdge == nil && sNodeBinary == nil { // sNode is a binary leaf
-			break
+		elementKey := FeltToKey(globalTrieHeight, keys[0])
+		if !firstKey.Equal(&elementKey) {
+			return false, errors.New("correct proof but invalid key")
 		}
-		parentKey = nodesFromRoot[i].key
-	}
-	return proofNodes, nil
-}
 
-// verifyProof checks if `leafPath` leads from `root` to `leafHash` along the `proofNodes`
-// https://github.com/eqlabs/pathfinder/blob/main/crates/merkle-tree/src/tree.rs#L2006
-func VerifyProof(root *felt.Felt, key *Key, value *felt.Felt, proofs []ProofNode, hash hashFunc) bool {
-	expectedHash := root
-	remainingPath := NewKey(key.len, key.bitset[:])
-	for i, proofNode := range proofs {
-		if !proofNode.Hash(hash).Equal(expectedHash) {
-			return false
+		if val == nil || !values[0].Equal(val) {
+			return false, errors.New("correct proof but invalid value")
 		}
 
-		switch proofNode := proofNode.(type) {
-		case *Binary:
-			if remainingPath.Test(remainingPath.Len() - 1) {
-				expectedHash = proofNode.RightHash
-			} else {
-				expectedHash = proofNode.LeftHash
-			}
-			remainingPath.RemoveLastBit()
-		case *Edge:
-			subKey, err := remainingPath.SubKey(proofNode.Path.Len())
-			if err != nil {
-				return false
-			}
-
-			// Todo:
-			// If we are verifying the key doesn't exist, then we should
-			// update subKey to point in the other direction
-			if value == nil && i == len(proofs)-1 {
-				return true
-			}
-
-			if !proofNode.Path.Equal(subKey) {
-				return false
-			}
-			expectedHash = proofNode.Child
-			remainingPath.Truncate(251 - proofNode.Path.Len()) //nolint:mnd
-		}
+		return hasRightElement(rootKey, &firstKey, nodes), nil
 	}
 
-	return expectedHash.Equal(value)
-}
-
-// VerifyRangeProof verifies the range proof for the given range of keys.
-// This is achieved by constructing a trie from the boundary proofs, and the supplied key-values.
-// If the root of the reconstructed trie matches the supplied root, then the verification passes.
-// If the trie is constructed incorrectly then the root will have an incorrect key(len,path), and value,
-// and therefore it's hash won't match the expected root.
-// ref: https://github.com/ethereum/go-ethereum/blob/v1.14.3/trie/proof.go#L484
-func VerifyRangeProof(root *felt.Felt, keys, values []*felt.Felt, proofKeys [2]*Key, proofValues [2]*felt.Felt,
-	proofs [2][]ProofNode, hash hashFunc,
-) (bool, error) {
-	// Step 0: checks
-	if len(keys) != len(values) {
-		return false, fmt.Errorf("inconsistent proof data, number of keys: %d, number of values: %d", len(keys), len(values))
+	// In all other cases, we require two edge paths available.
+	// First, ensure that the last key is greater than the first key
+	if last.Cmp(first) <= 0 {
```

Comment:
Redundant, impossible case. Please notice we have `monotonically increasing` check.
...or I totally missed what is `first`

Reply:
As mentioned in the previous comment, `first` may or may not equal to `key[0]`, so the monotonically increasing check doesn't apply here.